docs/style_guide.md

```diff
@@ -251,3 +251,27 @@ end
 
 `negate` is the version of the function that runs in OCaml, and `Checked.negate` is the one that runs inside of a SNARK circuit.
 
+### Polymorphic variants
+
+OCaml allows so-called polymorphic variants, with constructors marked with a backtick:
+
+```ocaml
+type truth = [ `True | `False ]
+```
+
+One advantage of such variants over ordinary algebraic datatypes is that they allow easy extensibility:
+
+```ocaml
+type uncertainty = [ truth | `Maybe ]
+```
+
+The constructors for polymorphic variants live in a global namespace, and they're not associated a unique defined type. One result is that the OCaml compiler can't check for exhaustive pattern-matching when using polymorphic variants. Their extensibility means both of these statements are correct:
```

Comment:
> One result is that the OCaml compiler can't check for exhaustive pattern-matching when using polymorphic variants.

Can you provide an example where this is true? I'm not familiar with this. AFAIK, OCaml can check for  exhaustive pattern-matching. It either checks against the polymorphic variant set it is aware of, or if it doesn't know the type, it infers a closed polymorphic variant set from the match statement.

Reply:
I could define
```
let f = function `True -> ... | `False -> ...
```
which I mean to apply to values of type `uncertainty`. If a value is `True` but not known to be of that type (it could be of type `truth`), then the compiler accepts applying `f` to that value without any warning. If I annotate the type for `f`, or for the value, yes, I get the exhaustiveness warning.

RW OCaml has an example where a pattern match includes a variant not mentioned in a defined type, but doesn't lead to a warning.

Maybe I should reword the statement to say: Pattern-matching with polymorphic variants is subtler than when matching against ordinary datatype constructors, perhaps leading to unexpected behavior.

Or, I could drop it ...

Reply:
Not sure I see how this is a bad thing. It doesn't let you pass in values for cases that aren't matched, it just let's you pass in subtypes, right? I think that's actually advantageous in certain scenarios. I'm not saying we shouldn't have any rules about when to use them, but I just think disallowing this pattern entirely is not the right rule, as it doesn't introduce any errors. I can't think of a scenario off my head where I would cause a bug because of this, but perhaps there is one.

Also, I don't think polymorphic variants should be the default, just an option, and we should explicitly state the cases and reasons it is valid to use, and otherwise fallback to old school data types.